### ISSUE TITLE
EPIC v2 multiple `IlmnID` per CpG `Name` behavior

### ISSUE THREAD

**Issue by hhp94** [perishky/meffil#48]
Dear Dr. Suderman,

Thank you so much for adding EPIC v2 support to `{meffil}` so quickly.

[It looks like](https://support.illumina.com/array/array_kits/infinium-methylationepic-beadchip-kit/downloads.html) there are now multiple `IlmnID` for a cpg `Name` in the manifest. Because of this, the CpG name in the beta matrix returned by `{meffil}` is no longer unique as in EPIC v1. For example, `cg00002033` is replicated twice. In the manifest, the `IlmnID` would be 
`cg00002033_TC11` and `cg00002033_TC12` and the `Name` column would be 2 values of `cg00002033`.

I was just wondering how this duplication is handled by `{meffil}`? QCing my data with the current version of `{meffil}`, it looks like the beta values for a duplicated cpg have the same exact value. For example, `cg00002033` is duplicated twice. `cg00002033(1)` and `cg00002033(2)` are the same for each individual in my data. 
![screen1](https://user-images.githubusercontent.com/58577711/229825677-2fc28acb-de83-453d-8504-5befab2f2765.png)

I'm not sure if this is an intended behavior since I would expect the values would be close but not the same. Even if we get back different values for each repeat, I'm not sure if averaging across the repeats is valid either. Do you have a recommendation?

Sincerely,
Hung

**Comment by perishky** [comment]
Hi Hung, 

There seem to be ~5000 duplicated probes.  You are correct, a duplicated probe should be similar but not exactly the same!  Thank you for pointing this out. I've now changed the behavior of meffil to something more sensible.  Each CpG site with duplicated probes will have the first probe named as usual and all remaining duplicates named using IlmnID values.  For example, you should now see probes named "cg00002033" and "cg00002033_TC12". They should be highly correlated but not identical.  There are a few options for EWAS interpretation.  The easiest is just to omit association testing for probes with IlmnID names (in R, just grep for "_").  These could be removed before or after the EWAS is performed. Alternatively, as you mentioned, you could replace the duplicates with the mean or median across all duplicates. My guess is that the resulting probe will have less technical noise.  A third alternative is to calculate the correlations between all pairs of duplicates and take the one with strongest overall correlation with the other duplicates.  My guess is that the best option is the second option with the median. I could add this an option (probably default) to the normalization pipeline if that would be useful.
Best,
Matt

**Comment by hhp94** [comment]
Dear Matt, 

Thank you for your answer. Your answer makes a lot of sense to me. May I ask a follow-up question regarding R?, I'm still relatively inexperienced.

Do you think this "dup handling" option should be a separate function? Or should it be part of `meffil.normalize.samples()`? Because a relatively clean strategy that I see is to:
* Let `meffil.normalize.samples()` return a matrix with `IlmnID` instead of `Name` as rownames. This helps subsetting easier.
* Write `meffil.EPICv2.format()` as something along the lines of the following

```r
library(data.table)
manifest # The manifest as a data.table object with just `IlmnID` and `Name`
str(manifest)
# Classes ‘data.table’ and 'data.frame':	937055 obs. of  2 variables:
#  $ IlmnID: chr  "cg25324105_BC11" "cg25383568_TC11" "cg25455143_BC11" "cg25459778_BC11" ...
#  $ Name  : chr  "cg25324105" "cg25383568" "cg25455143" "cg25459778" ...

# Filter to only Cpg that are dups
manifest_dup_only <- manifest[, N := .N, by = "Name"][N > 1]

# Most CpGs are duplicates, one has 10 replicates for some reason
unique(manifest_dup_only[, list(Name, N)])[, table(N)]
# N
#    2    3    4    5    6   10 
# 4188  962   23   48    3    1 

# Store the dup_only cpgs as an internal object because the size is small enough
manifest_dup_only$N <- NULL 
lobstr::obj_size(manifest_dup_only)
# 1.27 MB

str(manifest_dup_only)
# Classes ‘data.table’ and 'data.frame':	11622 obs. of  2 variables:
#  $ IlmnID: chr  "cg09085639_BC11" "cg15034763_TC11" "cg20651681_TC11" "cg20029347_TC11" ...
#  $ Name  : chr  "cg09085639" "cg15034763" "cg20651681" "cg20029347" ...

# This object holds all the unique _BC11, _TC11, etc. combination of suffixes of
# IlmnID. This is a safe way to handle the renaming of the beta matrix
all_unique_suffix <-
  tstrsplit(manifest$IlmnID, split = "_", fixed = TRUE,keep = 2)[[1]] |> 
  unique()

# Possibly also an internal object
suffix_grep <- paste(paste0("_", all_unique_suffix), collapse = "|")
all_unique_suffix |> head()
# [1] "BC11" "TC11" "BO11" "TO11" "BC12" "TC12"
suffix_grep 
# [1] "_BC11|_TC11|_BO11|_TO11|_BC12|_TC12|

#' Format EPICv2 beta matrix
#' 
#' Convert the output of [meffil.normalize.samples()] using EPICv2 chip into 
#' typical format with cpg names as row names.
#'
#' @inheritParams meffil.normalize.quantiles 
#' @param method which method to handle repeated cpg? Default to "median"
#'
#' @return a beta matrix with cpg `Name` as row names
#' @export
meffil.EPICv2.format <- function(beta, method = c("median", "mean", "cor")) {
  require(data.table)
  # Assuming beta is not a GDS filename
  method <- match.arg(method)
  if(method == "median") {
    ag_fns <- median # ag stands for aggregate
  }
  
  sample_names <- colnames(beta)
  
  # Convert to data.table because we are only working with ~11k CpGs
  ## get just the dupped CpGs from the input beta matrix using the internal
  ## object manifest_dup_only
  dups_dt <- beta[manifest_dup_only$IlmnID, ] |>
    as.data.table(keep.rownames = "IlmnID") |> # Convert to data.table, keep row names as "IlmnID"
    merge(manifest_dup_only, by = "IlmnID") # Merge with the manifest_dup_only again to get "Name"
  
  # Do column wise aggregation. 
  ag_dt <- dups_dt[, lapply(.SD, ag_fns, na.rm = TRUE), .SDcols = sample_names, by = "Name"] |>
    as.matrix(rownames = "Name") # Convert back to matrix. This is the dupped part of beta
  
  # Renaming beta matrix
  ### Removing all the suffixes using the `suffix_grep` object 
  row.names(beta) <- sub(pattern = suffix_grep, replacement = "", row.names(beta))
  return(
    rbind(
      beta[!row.names(beta) %in% row.names(ag_dt), ], # The part without the dups
      ag_dt # The aggregated dups
      )
    )
}
```

Sincerely,
Hung

**Comment by perishky** [comment]
Hi Hung, 

I spent a bit of time trying out different solutions and have come up with the following (which is quite similar to what you suggested, thank you!):

- meffil.normalize.samples() has a new argument "dup.fun" which is used to 'collapse' duplicate probes. The default is 'median' but any function that takes a vector of numbers and returns a single number can be used.  If "dup.fun" is NULL, then duplicates are not collapsed. 
- a new function meffil.collapse.dups() can be used to collapse duplicate probes if they were not collapsed by meffil.normalize.samples().   It has two arguments, methylation matrix and the function "dup.fun" for collapsing duplicates.  The default is 'median'.

Hopefully that makes sense!  Do let me know if you think this could be improved or if you notice a problem.

Best,
Matt

**Comment by hhp94** [comment]
This is great. Thank you so much Matt. I will test out the new function with my data and get back to you if I find any other issues. 

Sincerely,
Hung

**Comment by mikaelahu** [comment]
Dear Dr Suderman, 

I am not sure whether I should have started a new issue, but since I found the above conversation helpful I decided to continue here. Firstly, I greatly appreciate that meffil.normalize.samples collapses duplicate values to median. However, now that there is only one row per beta, I am not sure how to know which annotation row to use from the "EPICv2.hg38.manifest.tsv". 

As far as I see, I could 1) not collapse the replicates, using meffil.normalize.samples(dup.fun=NULL), select the median value for each probe myself and store the full probeID (containing the prefixes_suffix) for later, or 2) use, for instance, the mean of the replicate values, which I would do with meffil.normalize.samples(dup.fun = function(x) mean(x, na.rm = T)) if I have understood correctly. 

However I was wondering, is there any way within meffil to know to which row of the annotation file the collapsed probe values correspond to? 

Thank you already in advance,
Mikaela
